File: ruby/linux.rb
class Linux
	def initialize(cpu, mem)
		@cpu = cpu
		@mem = mem
	end
	
	def syscall_return_int(value)
		Pointer.new(@cpu.register[0], 0, 8).write_int value
	end

	def handle_syscall(number, args)
		case number
		when 1 # write
			fd = args[0]
			pos = args[1]
			size = args[2]
			if fd == 1 # stdout
				data = @mem.read(pos, size)
				print data.pack("C*")
				syscall_return_int size
			else
				raise "not implemented fd: %d" % fd
			end
		when 60 # exit
			puts "exit code: %d" % args[0]
			@cpu.stopped = true
		when 102, 104, 107, 108 # getuid, getgid, geteuid, getegid
			syscall_return_int 1000 # default UID of the first user on Ubuntu
		else
			raise "syscall not implemented: %d (0x%x)" % [number, number]
		end
	end
end

File: python/linux.py
import struct
from pointer import Pointer


class Linux:
    def __init__(self, cpu, mem):
        self.cpu = cpu
        self.mem = mem

    def syscall_return_int(self, value):
        Pointer(self.cpu.register[0], 0, 8).write_int(value)

    def handle_syscall(self, number, args):
        if number == 1:  # write
            fd = args[0]
            pos = args[1]
            size = args[2]
            if fd == 1:  # stdout
                data = self.mem.read(pos, size)
                print(struct.pack('B' * size, data))
                self.syscall_return_int(size)
            else:
                raise "not implemented fd: %d" % fd
        elif number == 12:  # brk
            self.syscall_return_int(0)
        elif number == 60:  # exit
            print("exit code: %d" % args[0])
            self.cpu.stopped = True

        # getuid, getgid, geteuid, getegid
        elif number in [102, 104, 107, 108]:
            # 1000 is default UID of the first user on Ubuntu
            self.syscall_return_int(1000)
        else:
            raise "syscall not implemented: %d (0x%x)" % [number, number]
